### PR TITLE
feat: use cached result while onboarding repos

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Add a new interface OnboardingBranchCache
- cache isOnboarded result and use it to prevent cloning when parent and onboarding-branch haven't been updated between runs
- check if repoIsOnboarded before calling `detectRepoFileConfig`, don't call if not onboarded

## Context
- prevent cloning while onboardingRepos

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/git/onboarding-branch-cache.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import type { OnboardingBranchCache } from '../cache/repository/types';
+
+export function getCachedOnboardingBranch(
+  branchSha: string,
+  baseBranchSha: string,
+  baseBranchName: string
+): OnboardingBranchCache | null {
+  const cache = getCache();
+  const { onboardingBranch } = cache;
+  if (!onboardingBranch || onboardingBranch?.sha === null) {
```
Comment (viceice) [inline lib/util/git/onboarding-branch-cache.ts:12]:
```suggestion
  if (!onboardingBranch || onboardingBranch.sha === null) {
```
can the `sha` be undefined?

File: lib/util/git/onboarding-branch-cache.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    return null;
+  }
+
+  onboardingBranch.parentSha ??= cache.scan?.[baseBranchName].sha;
+  if (
+    onboardingBranch?.sha !== branchSha ||
```
Comment (viceice) [inline lib/util/git/onboarding-branch-cache.ts:18]:
```suggestion
    onboardingBranch.sha !== branchSha ||
```

File: lib/util/git/onboarding-branch-cache.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+export function setOnboardingBranchCache(
+  branchSha: string,
+  baseBranchSha: string,
+  isOnboarded: boolean,
+  extractedDependencies?: ExtractResult
+): void {
+  const cache = getCache();
+  const { onboardingBranch = {} as OnboardingBranchCache } = cache;
```
Comment (viceice) [inline lib/util/git/onboarding-branch-cache.ts:33]: are you sure we need the cast? if yes:

```suggestion
  const onboardingBranch = cache.onboardingBranch ?? {};
```

File: lib/workers/repository/init/merge.ts [hunk 4]
```diff
@@ -27,12 +27,16 @@ export async function detectRepoFileConfig(): Promise<RepoFileConfig> {
...
+      let configFileParsed = (await platform.getJsonFile(configFileName))!;
+      if (configFileParsed) {
+        if (configFileName === 'package.json') {
+          configFileParsed = configFileParsed.renovate;
+        }
+        return { configFileName, configFileParsed };
       }
-      return { configFileName, configFileParsed };
+    } catch (err) {
+      // probably file doesn't exist
```
Comment (viceice) [inline lib/workers/repository/init/merge.ts:39]: i think this can't really happen. the only scenario is:

- renovate fetches Head Sha
- user deletes renovate config
- renovate ties to fetch config file

in this case throw a repo changed error

File: lib/workers/repository/onboarding/branch/index.ts [hunk 5]
```diff
@@ -21,7 +29,27 @@ export async function checkOnboardingBranch(
   logger.debug('checkOnboarding()');
   logger.trace({ config });
   let onboardingBranch = config.onboardingBranch;
-  const repoIsOnboarded = await isOnboarded(config);
+  let repoIsOnboarded: boolean;
+  const onboardingBranchSha = getBranchCommit(onboardingBranch!);
+  const baseBranchSha = getBranchCommit(config.baseBranch!);
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.ts:34]: i think this can only be default branch 
```suggestion
  const baseBranchSha = getBranchCommit(config.defaultBranch!);
```

File: lib/workers/repository/onboarding/branch/index.ts [hunk 6]
```diff
@@ -21,7 +29,27 @@ export async function checkOnboardingBranch(
...
+  // TODO #7154
+  const cachedOnboardingBranch = getCachedOnboardingBranch(
+    onboardingBranchSha!,
+    baseBranchSha!,
+    config.baseBranch!
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.ts:39]: same here

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/git/onboarding-branch-cache.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/util/git/onboarding-branch-cache.ts:12]: For the initial run it could be.

File: lib/workers/repository/init/merge.ts (hunk 4 above)
Comment (RahulGautamSingh) [reply lib/workers/repository/init/merge.ts:39]: Happens during tests when repo not onboarded, we have same logic inside the isOnboarded function which has a try-catch block to prevent erros.

File: lib/util/git/onboarding-branch-cache.ts (hunk 3 above)
Comment (RahulGautamSingh) [reply lib/util/git/onboarding-branch-cache.ts:33]: onboardingBranch is optional property so without type-cast then TS will throw error.

File: lib/util/cache/repository/types.ts [hunk 7]
```diff
@@ -32,6 +33,13 @@ export interface BranchCache {
   upgrades: BranchUpgradeCache[];
 }
 
+export interface OnboardingBranchCache {
+  isOnboarded: boolean;
+  parentSha?: string;
+  sha: string;
```
Comment (rarkins) [inline lib/util/cache/repository/types.ts:39]:
```suggestion
  defaultBranchSha?: string;
  onboardingBranchSha: string;
```

More easy to understand names.

Also, why is sha mandatory but parentSha optional currently?

File: lib/util/cache/repository/types.ts [hunk 8]
```diff
@@ -44,6 +52,7 @@ export interface RepoCacheData {
   };
   gitConflicts?: GitConflictsCache;
   prComments?: Record<number, Record<string, string>>;
+  onboardingBranch?: OnboardingBranchCache;
```
Comment (rarkins) [inline lib/util/cache/repository/types.ts:55]:
```suggestion
  onboarding?: OnboardingBranchCache;
```

File: lib/util/git/onboarding-branch-cache.ts (hunk 1 above)
Comment (rarkins) [reply lib/util/git/onboarding-branch-cache.ts]: Shouldn't the object be absent in the initial run?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/cache/repository/types.ts (hunk 7 above)
Comment (RahulGautamSingh) [reply lib/util/cache/repository/types.ts:39]: parentSha is stored in the cache.scan object so I was thinking to use that in future instead of storing the same thing twice.

File: lib/util/git/onboarding-branch-cache.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/util/git/onboarding-branch-cache.ts]: Yes if its undeinfed we return null, incase if the branch is defined but sha is null we return null too. There shouldn't be any case where the branch is defined and sha is null . But all current methods which return branchCommit have return type `string|null` so had to add thiss check due to TS errors.

File: lib/util/git/onboarding-branch-cache.ts [hunk 9]
```diff
@@ -1,3 +1,4 @@
+// istanbul ignore file
```
Comment (rarkins) [inline lib/util/git/onboarding-branch-cache.ts:1]:
```suggestion
```

Comment (RahulGautamSingh) [reply lib/util/git/onboarding-branch-cache.ts:1]: Added this to prevent the codecov errors, will remove this when I add tests for the added file.

File: lib/workers/repository/init/merge.ts [hunk 10]
```diff
@@ -173,7 +177,7 @@ export async function mergeRenovateConfig(
 ): Promise<RenovateConfig> {
   let returnConfig = { ...config };
   let repoConfig: RepoFileConfig = {};
-  if (config.requireConfig !== 'ignored') {
+  if (config.repoIsOnboarded && config.requireConfig !== 'ignored') {
```
Comment (RahulGautamSingh) [inline lib/workers/repository/init/merge.ts:180]: Any comments on this change?

File: lib/util/cache/repository/types.ts [hunk 11]
```diff
@@ -32,6 +33,13 @@ export interface BranchCache {
   upgrades: BranchUpgradeCache[];
 }
 
+export interface OnboardingBranchCache {
+  isOnboarded: boolean;
```
Comment (rarkins) [inline lib/util/cache/repository/types.ts:37]: Will this object also be stored if isOnboarded is `false`, or only for onboarding repos?

File: lib/util/cache/repository/types.ts [hunk 12]
```diff
@@ -32,6 +33,13 @@ export interface BranchCache {
   upgrades: BranchUpgradeCache[];
 }
 
+export interface OnboardingBranchCache {
```
Comment (rarkins) [inline lib/util/cache/repository/types.ts:36]:
```suggestion
export interface OnboardingCache {
```

File: lib/workers/repository/init/merge.ts (hunk 10 above)
Comment (rarkins) [reply lib/workers/repository/init/merge.ts:180]: Don't we merge renovate config from the onboarding branch?

If that scenario doesn't touch this then are we better to avoid calling this function altogether in cases where repo isn't onboarded?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/cache/repository/types.ts (hunk 11 above)
Comment (RahulGautamSingh) [reply lib/util/cache/repository/types.ts:37]: Yes, our main use-case is when isOnboarded is `false` to prevent the cloning on re-runs after the onboardingPR is created.

Comment (rarkins) [reply lib/util/cache/repository/types.ts:37]: So then is this object is present then `isOnboarded` will always be `false`? Maybe redundant then?

File: lib/workers/repository/init/merge.ts (hunk 10 above)
Comment (RahulGautamSingh) [reply lib/workers/repository/init/merge.ts:180]: Yes we do. We can bypass this for PRs which are not onboarded by storing the onboardingConfig in cache and use it when we don't clone.

Comment (rarkins) [reply lib/workers/repository/init/merge.ts:180]: Just beware that this would have a huge negative impact if we overlook something and get it wrong

File: lib/util/cache/repository/types.ts (hunk 11 above)
Comment (RahulGautamSingh) [reply lib/util/cache/repository/types.ts:37]: Yes I will remove it once I add the code to delete the cachedOnboardingBranch.

File: lib/workers/repository/init/merge.ts [hunk 13]
```diff
@@ -49,12 +49,16 @@ export async function detectRepoFileConfig(): Promise<RepoFileConfig> {
   const cache = getCache();
   let { configFileName } = cache;
   if (configFileName) {
-    let configFileParsed = (await platform.getJsonFile(configFileName))!;
-    if (configFileParsed) {
-      if (configFileName === 'package.json') {
-        configFileParsed = configFileParsed.renovate;
+    try {
+      let configFileParsed = (await platform.getJsonFile(configFileName))!;
```
Comment (viceice) [inline lib/workers/repository/init/merge.ts:53]:
```suggestion
      let configFileParsed = await platform.getJsonFile(configFileName);
```
is this related to this PR? should we extract to separate PR?

File: lib/workers/repository/onboarding/branch/index.ts [hunk 14]
```diff
@@ -21,7 +29,27 @@ export async function checkOnboardingBranch(
...
+  const baseBranchSha = getBranchCommit(config.defaultBranch!);
+  // TODO #7154
+  const cachedOnboardingBranch = getCachedOnboardingBranch(
+    onboardingBranchSha!,
+    baseBranchSha!,
+    config.defaultBranch!
+  );
+  if (cachedOnboardingBranch === null) {
+    repoIsOnboarded = await isOnboarded(config);
+  } /* istanbul ignore next*/ else {
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.ts:43]: can't we add a test instead?

File: lib/workers/repository/init/merge.ts (hunk 13 above)
Comment (RahulGautamSingh) [reply lib/workers/repository/init/merge.ts]: Can be extracted to another PR 👍
